enhancements/machine-api/machine-deletion-hooks.md

```diff
@@ -0,0 +1,525 @@
+
+The Machine status will contain conditions identifying why the Machine controller
+has not progressed in removing the Machine. This will be in the form of new `Drainable`
+and `Terminable` conditions.
+
+It is expected that hook implementing controllers should signal to users when they
+are having complete system issues via some mechanism (eg. their `ClusterOperator` status if the issue is endangering
+the cluster health) and that they should also use conditions on the Machine to
+identify issues with removing the hook from a particular Machine.
+
+When errors occur, hook implementing controllers are encouraged to add a condition
+to Machines with details of the error. The condition `Type` should be formad as the
+hook name, prefixed by the lifecycle point. For example, for a `preDrain` hook with name
+`MigrateImportantApp`, the condition `Type` would be `PreDrainMigrateImportantApp`.
+
+For example, when an error has occured in a hook implementing controller with a `preDrain` hook.
+The `Drainable` condition will contain the name and owner of the hook blocking the removal.
+A user should then also expect to see another condition named based on the pattern described above.
+The user may also look towards the owner's own reporting to identify issues for wider system issues.
+
+We also encourage hook implementing controllers to use Event objects to add detail
+about the operations they are performing. For example if the hook implementing
+controller is waiting on some side effect, it should periodically send an Event,
+attached to the Machine object, notifying the user of delay. Events are already
+displayed in the OpenShift console and in `oc describe` and as such should be
+relatively easy for the end user to find.
+
+##### Hook ordering
+
+The machine controller will not attempt to enforce any ordering of hooks.
+No ordering should be expected by the machine controller.
+
+Hook Implementing Controllers may choose to provide a mechanism to allow
+ordering amongst themselves via whatever means HICs determine.
+Examples could be using CRDs external to the Machine API, gRPC communications,
+or additional annotations on the Machine or other objects.
+
+#### Hook Implementing Controller Design
+
+Hook Implementing Controller is the component that manages a particular
+lifecycle hook.
+
+##### Hook Implementing Controllers must
+
+* Watch Machine objects and determine when an appropriate action must be taken.
+* After completing the desired hook action, remove the hook.
+
+##### Hook Implementing Controllers may
+
+* Watch Machine objects and add a hook as desired by the cluster administrator.
+* Coordinate with other Hook Implementing Controllers through any means
+possible, such as using common annotations, CRDs, etc. For example, one hook
+controller could set an annotation indicating it has finished its work, and
+another hook controller could wait for the presence of the annotation before
```

Comment:
still referencing annotations here

Reply:
I think this one is a legitimate reference to an annotation, WDYT?

Reply:
yes, sorry, my confusion